### PR TITLE
[FLINK-21580] Remove ScheduleMode

### PR DESCRIPTION
**Pull request by tillrohrmann**
This PR removes the `ScheduleMode` from the `JobGraph` and the `ExecutionGraph`.

* 1fb9313 [FLINK-21580] Introduce TaskDeploymentDescriptorFactory.PartitionLocationConstraint:
  TaskDeploymentDescriptorFactory.PartitionLocationConstraint replaces the boolean allowUnknownPartitions in
the TaskDeploymentDescriptorFactory.

* 73fdca1 [FLINK-21580] Remove ScheduleMode from ExecutionGraph 
  Removes ScheduleMode from ExecutionGraph and replace it with TaskDeploymentDescriptorFactory.PartitionLocationConstraint.

* b331a23 [FLINK-21580] Remove ScheduleMode from JobGraph and its test
  The JobGraph's JobType replaces the ScheduleMode. Before we always configured the
ScheduleMode in accordance with the JobType. Now, only the JobType defines which
type of scheduling is selected.
   
    JobType.STREAMING: The whole job must be deployed at the same time
    JobType.BATCH: The job can be executed in steps

### CODE REVIEW

File: flink-tests/src/test/java/org/apache/flink/test/runtime/BlockingShuffleITCase.java [hunk 1]
```diff
@@ -94,7 +93,6 @@ private JobGraph createJobGraph(int numRecordsToSend) {
 
         StreamGraph streamGraph = env.getStreamGraph();
         streamGraph.setGlobalDataExchangeMode(GlobalDataExchangeMode.ALL_EDGES_BLOCKING);
-        streamGraph.setScheduleMode(ScheduleMode.LAZY_FROM_SOURCES);
         // a scheduler supporting batch jobs is required for this job graph, because it contains
         // blocking data exchanges, and the "lazy from sources" schedule mode is not supported by
```
Comment (rmetzger) [inline flink-tests/src/test/java/org/apache/flink/test/runtime/BlockingShuffleITCase.java:97]: comment still mentions the schedule mode

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGeneratorTest.java [hunk 2]
```diff
@@ -853,9 +852,9 @@ public void testIteration() {
...
     }
 
-    /** Test default schedule mode. */
+    /** Test default job type. */
     @Test
-    public void testDefaultScheduleMode() {
+    public void testDefaultJobType() {
         StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
 
         // use eager schedule mode by default
```
Comment (rmetzger) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGeneratorTest.java:860]: comment still refers to schedule mode

**Review by rmetzger — APPROVED** [review]
Thanks a lot for the cleanup! +1 to merge once CI is green.
I had two minor comments.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java [hunk 3]
```diff
@@ -304,7 +300,7 @@ public ExecutionGraph(
 
         this.blobWriter = Preconditions.checkNotNull(blobWriter);
 
-        this.scheduleMode = checkNotNull(scheduleMode);
+        this.partitionLocationConstraint = partitionLocationConstraint;
```
Comment (rmetzger) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java:303]: nit: checkNotNull() ?

File: flink-tests/src/test/java/org/apache/flink/test/runtime/BlockingShuffleITCase.java (hunk 1 above)
Comment (tillrohrmann) [reply flink-tests/src/test/java/org/apache/flink/test/runtime/BlockingShuffleITCase.java:97]: True. I will correct it. Good catch.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGeneratorTest.java (hunk 2 above)
Comment (tillrohrmann) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGeneratorTest.java:860]: True. Good catch. Will correct it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java (hunk 3 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java:303]: Yeah, let's do it for reasons of symmetry.

**Comment by tillrohrmann** [comment]
Thanks for the review @rmetzger. I will address your comments and then merge this PR.

**Pull request closed without merging**